aodndata/aatams_sattag_qc_dm/csv_schema.py

```diff
@@ -0,0 +1,279 @@
+          keys():
+
+          keys: the selected keys.
+
+        Returns:
+
+        """
+        cross_values = {}
+        if set(keys).issubset(datadict.keys()):
+            cross_values = {x: set() for x in keys}
+            _ = [cross_values[x].update(datadict[x]) for x in cross_values.keys()]
+        return cross_values
+
+    @classmethod
+    def report(cls, msg):
+        """Report a msg.
+
+        Args:
+          msg(str): the msg
+
+        Returns:
+
+        """
+        print(cls.__name__ + ": " + msg)
+
+    def validate_file(self, file):
+        """Validate a csv file.
+
+        Args:
+          file(str): a csv file
+
+        Returns:
+          header(Dict[str,List[Any]]): A dict with header columns converted as keys
+          data(Dict[str,List[Any]]): A dict with data columns converted as keys
+
+        """
+        self.report("Validation Started for %s" % file)
+        self.report("\tLoading content schema for %s" % file)
+
+        schema_name = self.file2schema(file)
+        fschema = self.file_schemas[schema_name]
+
+        self.report("\tLoading data from %s" % file)
+        _, header, data_list = self.load_csv(
+            file, bulk_load=self.bulk_load, skip_every=self.skip_every,
+        )
+
+        self.report("\tValidating Header in %s" % file)
+        valid_header = self.validate_header(header, fschema)
+
+        self.report("\tAggregating data rows for validation in %s" % file)
+        col_dict = self.aggregate_to_dict(header, data_list)
+
+        self.report("\tValidating data types in %s" % file)
+        valid_dataset = {}
+        for key, value in col_dict.items():
+            # TODO continue to validate per column since the fail output can be verbose!?
+            valid_dataset[key] = Schema([fschema[key]]).validate(value)
+            self.report("\t\t %s[%s] is valid" % (file, key))
+
+        if self.same_keys_content:
+            self.report("\t\t\tValidating inter columns in %s" % file)
+            self.validate_self_content(self.same_keys_content, valid_dataset)
+
+        if self.cross_content_scope and self.cross_values:
+            self.report("\t\t\t\tValidating cross columns references in %s" % file)
+            for key, value in self.cross_values.items():
+                isdiff = set(valid_dataset[key]).difference(value)
+                if isdiff:
+                    errmsg = (
+                        "Cross content comparison failed.\n\
+                            %s[%s] = %s\n~\n%s"
+                        % (file, key, value, self.cross_values[key])
+                    )
+                    raise SchemaError(errmsg)
+        self.report("Validation ended for %s" % file)
+
+        return valid_header, valid_dataset
+
+    def __init__(self, bulk_load=True, skip_every=0):
+        """Initialize the CSV schema Class."""
+        self.bulk_load = bulk_load
+        self.skip_every = 0 if skip_every < 0 else skip_every
```

Comment:
Correct me please if I'm wrong, but I don't understand the need for ```skip_every``` throughout your different functions. It seems like your value is always set to 0, but you create different conditions (such as in  ```load_csv```) depending on this value.

Also more comments would greatly help to explain some of the init values (and everything in general). It can be quite hard to follow for my modest sapiens brain :D

Reply:
This is the CSVSchema class, and it was created as a separated class to be re-used :).  You can use ```skip_every``` if you define so at init time on the SCHEMA_CLASS init method.

yeap - I was lazy in defining the init method of the SCHEMA_CLASS to be ```*args,**kwargs```. I will expand that on to match the CSVSchema init signature.

Reply:
> This is the CSVSchema class, and it was created as a separated class to be re-used :). You can use `skip_every` if you define so at init time on the SCHEMA_CLASS init method.

in that case consider putting everything you think could be re-used into python-aodncore

Reply:
see with @lwgordonimos for some details if needed

Reply:
I think code that gets this deep into the data format is better in aodndata (or aodntools?), and can be shared by putting it in the common module e.g. https://github.com/aodn/python-aodndata/tree/master/aodndata/common

There's nothing wrong with this sort of thing to share code within this package, depending on whether it's facility specific or more generally useful:

`from aodndata.common import CSVSchema`
`from aodndata.aatams import CSVSchema`

I think aodncore is better left for lower level core things, like defining the abstract workflow and very generic utility code.

Reply:
>in that case consider putting everything you think could be re-used into python-aodncore

Yeap, but I put it here to see if the feature will be of interest or not.

@lwgordonimos - OK - I will leave in here at ```aodndata/common``` since it's a data thing and very optional at the moment.